delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/files/FormDataExtractor.scala

```diff
@@ -41,55 +43,82 @@ object FormDataExtractor {
+        createStreamed = (_, parts) => Multipart.FormData(parts),
+        createStrictBodyPart =
+          (entity, headers) => Multipart.General.BodyPart.Strict(entity, headers).toFormDataBodyPart.get,
+        createStrict = (_, parts) => Multipart.FormData.Strict(parts)
+      )
+
+  def apply(
+      mediaTypeDetector: MediaTypeDetectorConfig
+  )(implicit uuidF: UUIDF, as: ActorSystem, sc: Scheduler): FormDataExtractor =
+    new FormDataExtractor {
+      override def apply(
+          id: Iri,
+          entity: HttpEntity,
+          maxFileSize: Long,
+          storageAvailableSpace: Option[Long]
+      ): IO[FileRejection, (FileDescription, BodyPartEntity)] = {
+        val sizeLimit = Math.min(storageAvailableSpace.getOrElse(Long.MaxValue), maxFileSize)
+        IO.deferFuture(um(entity.withSizeLimit(sizeLimit)))
+          .mapError {
+            case RejectionError(r)                  =>
+              WrappedAkkaRejection(r)
+            case Unmarshaller.NoContentException    =>
+              WrappedAkkaRejection(RequestEntityExpectedRejection)
+            case x: UnsupportedContentTypeException =>
+              WrappedAkkaRejection(UnsupportedRequestContentTypeRejection(x.supported, x.actualContentType))
+            case x: IllegalArgumentException        =>
+              WrappedAkkaRejection(ValidationRejection(Option(x.getMessage).getOrElse(""), Some(x)))
+            case x: ExceptionWithErrorInfo          =>
+              WrappedAkkaRejection(MalformedRequestContentRejection(x.info.format(withDetail = false), x))
+            case x                                  =>
+              WrappedAkkaRejection(MalformedRequestContentRejection(Option(x.getMessage).getOrElse(""), x))
+          }
+          .flatMap { formData =>
+            IO.fromFuture(
+              formData.parts
+                .mapAsync(parallelism = 1) {
+                  case part if part.name == fieldName =>
+                    val filename    = part.filename.getOrElse("file")
+                    val contentType = detectContentType(filename, part.entity.contentType)
+                    FileDescription(filename, contentType).runToFuture.map { desc =>
+                      Some(desc -> part.entity)
+                    }
+                  case part                           =>
+                    part.entity.discardBytes().future.as(None)
+                }
+                .collect { case Some(values) => values }
+                .toMat(Sink.headOption)(Keep.right)
+                .run()
+            ).mapError {
+              case _: EntityStreamSizeException =>
+                FileTooLarge(maxFileSize, storageAvailableSpace)
+              case th                           =>
+                WrappedAkkaRejection(MalformedRequestContentRejection(th.getMessage, th))
+            }.flatMap(IO.fromOption(_, InvalidMultipartFieldName(id)))
+          }
+      }
```

Comment:
I think your changes don't really affect this, but I think in general I look at a lot of code in this function and don't understand why we are doing a certain thing. I like your `detectContentType` function because it's describing the contents

Perhaps if you have a good understanding of this class right now there are other functions you could make